src/core/lib/resource_quota/memory_quota.cc

```diff
@@ -39,69 +43,106 @@ static constexpr size_t kMinReplenishBytes = 4096;
 
 ReclamationSweep::~ReclamationSweep() {
   if (memory_quota_ != nullptr) {
-    memory_quota_->FinishReclamation(sweep_token_);
+    memory_quota_->FinishReclamation(sweep_token_, std::move(waker_));
   }
 }
 
 //
 // ReclaimerQueue
 //
 
-const ReclaimerQueue::Index ReclaimerQueue::kInvalidIndex;
-
-void ReclaimerQueue::Insert(
-    std::shared_ptr<EventEngineMemoryAllocatorImpl> allocator,
-    ReclamationFunction reclaimer, Index* index) {
-  ReleasableMutexLock lock(&mu_);
-  if (*index < entries_.size() && entries_[*index].allocator == allocator) {
-    entries_[*index].reclaimer.swap(reclaimer);
-    lock.Release();
-    reclaimer({});
-    return;
+struct ReclaimerQueue::QueuedNode
+    : public MultiProducerSingleConsumerQueue::Node {
+  explicit QueuedNode(RefCountedPtr<Handle> reclaimer_handle)
+      : reclaimer_handle(std::move(reclaimer_handle)) {}
+  RefCountedPtr<Handle> reclaimer_handle;
+};
+
+struct ReclaimerQueue::State {
+  Mutex reader_mu;
+  MultiProducerSingleConsumerQueue queue;  // reader_mu must be held to pop
+  Waker waker ABSL_GUARDED_BY(reader_mu);
+
+  ~State() {
+    bool empty = false;
+    do {
+      delete static_cast<QueuedNode*>(queue.PopAndCheckEnd(&empty));
+    } while (!empty);
   }
-  if (free_entries_.empty()) {
-    *index = entries_.size();
-    entries_.emplace_back(std::move(allocator), std::move(reclaimer));
-  } else {
-    *index = free_entries_.back();
-    free_entries_.pop_back();
-    Entry& entry = entries_[*index];
-    entry.allocator = std::move(allocator);
-    entry.reclaimer = std::move(reclaimer);
+};
+
+void ReclaimerQueue::Handle::Orphan() {
+  if (auto* sweep = sweep_.exchange(nullptr, std::memory_order_relaxed)) {
+    sweep->RunAndDelete(absl::nullopt);
```

Comment:
We probably need acq+rel poair because `RunAndDelete` need the state of `sweep` right?

Reply:
done